### ISSUE TITLE
Tagging in the DB

### ISSUE THREAD

**Issue by lpatmo** [codebuddies/django-concept#16]
Pulled from a PR conversation: 

@BethanyG: 

Agreed. The classic way to model something like this would be with a many-to-many and a Tags table. This was more a quick nod to the API specs, as well as an experiment with postgres JSONB capabilities. So a bit of a hand-wave.

But as this system grows, I do have a few concerns/caveats with a "simple" many-to-many Tags table:

Tags will most likely be a freeform field on the front end. I don't really see our users choosing from an epic list of tags...and the "common convention" is for users to enter what they see fit. This means writing a bunch of code to search for already existing tags and similar tags, and only add truly new/unique tags to the DB. Even with that, the Tags table will potentially be a mess of overlapping and duplicate data from the get-go. Just mocking up 28 resources for testing data created a "mess" -- and I am only one human.

Learning Paths, Hangouts, Notes and Discussions will also be tagged, I would imagine. So there will be potentially a lot of relations tracing back to the Tags table, each having overlapping, but not necessarily identical topics....so each would have to access the code for searching for existing tags/adding new tags. With virtually everything connecting to the Tags table (and it being updated a lot per user, hangout, note, discussion, etc.), it could become a bottleneck fairly quickly.

Postgres can index JSONB, and perform many of the same queries on JSONB that it can on regular data with small syntax changes. Both autocomplete and recommendations can be done (albeit not at top speed) with documents. But yes - if we are going the RDBMS route, then we shouldn't use JSONB too often, or in too many places.

One option is to use https://github.com/jazzband/django-taggit along with https://github.com/glemmaPaul/django-taggit-serializer, and customize the code so that each model has it's tags. See this for a rundown and some thoughts: https://medium.com/sthzg/a-short-exploration-of-django-taggit-bb869ea5051f.

And all that being said, I don't like how the Tags field adds an extra level right now, so I'm going to look at re-writing the serializer to omit that, and do some more digging on how Tags are implemented in other systems to see some more options. Do you have a go-to model/resource you can share?? That would be great.

@brylie: 
We ended up using a lot of JSONB where I work, and it has been a bit of a wrestling match. The main take-away, from a senior developer colleague of mine, is to only use JSONB for specific optimizations and to default to using a relational model with Postgres.

As mentioned, we can use an auto-complete widget along with the django-taggit package to improve consistency and avoid bespoke tagging code.

----

To expand a little bit, JSONB is good to used when we need 'data locality', in effect avoiding a second query or JOIN to get a related entity.

In general, Tags are used to show related content, e.g. "resources tagged Python". I believe that it will be desirable to use these tags to create views showing similar items, resources, etc.

The performance penalty comes when we try to search within the JSONB column (even with an index) for a tag string in the name property, and then join related documents together.

For these reasons, JSONB does not seem like a good approach for tagging, particularly when the django taggit package makes a proper M2M tagging approach very straightforward.

Please excuse my directness, but I strongly recommend we use the django-taggit approach here or to omit the tagging feature from this pull request.

**Comment by brylie** [comment]
@lpatmo thanks for carrying this discussion forward as a separate issue. @BethanyG thank you for your excellent work on the DRF prototype, and for prompting the interesting inquiry into data modelling using JSONB in a relational context :-)

These are the main reasons I believe we should use a many-to-many relationship for tags:
- many-to-many for taxonomy is a conventional and well-paved path
- separation of concerns between the *data model* and serialized/nested *API response data*
- a JSONB *array* may introduce unnecessary development complexity and resource usage

I have also included parts of the [Zen of Python](https://www.python.org/dev/peps/pep-0020/), where relevant, as a friendly-spirited nod to general Pythonic principles.

## Relationships
This is based on the idea of **using tags as a taxonomy to find related items**, which don't have to be of the same type. For example, when viewing a scheduled hangout, it might be useful to see both related hangouts, code buddies, and educational resources.

## Separation of concerns
> Flat is better than nested.

While we are conforming to an API specification with nested JSON output, that does not mean our data model needs to use nesting. The API can serialize nested documents while having a many-to-many relationship in the data model.

## Simplicity
> Simple is better than complex.

Aligning with a relational model, and using the Django Taggit library, will give us *simple* and relatively *clear* developer semantics directly from the Django ORM.

The Taggit library abstracts away code for managing tags and preventing duplicates, across all entities that use tags. Django querysets can be entity specific, while still using a common vocabulary of tags.

## Developer experience
> Readability counts.
> If the implementation is hard to explain, it's a bad idea.
> If the implementation is easy to explain, it may be a good idea.

It is not clear to me how the Django ORM, or even resorting to SQL queries, allows us to clearly query an *array* of JSON data. In effect, we don't simply have a nested document with key value pairs, but rather an array of nested documents with key/value pairs. 

The Postgres documentation shows containment operators for finding rows where a [top-level nested entity contains a key/value](https://www.postgresql.org/docs/current/functions-json.html#FUNCTIONS-JSONB-OP-TABLE). Additionally, [PostgreSQL JSON documentation](https://www.postgresql.org/docs/9.4/datatype-json.html#JSON-DOC-DESIGN) seems to indicate that storing an *array* of `tag` entities might not be advisable:

> Ideally, JSON documents should each represent an atomic datum that business rules dictate cannot reasonably be further subdivided into smaller datums that could be modified independently.

Django ORM [JSONField query examples](https://docs.djangoproject.com/en/2.2/ref/contrib/postgres/fields/#querying-jsonfield) show a single 'entity' [contained in](https://docs.djangoproject.com/en/2.2/ref/contrib/postgres/fields/#std:fieldlookup-hstorefield.contains) a parent row or having certain [keys or values](https://docs.djangoproject.com/en/2.2/ref/contrib/postgres/fields/#has-key).

Concretely, what would the query/code look like, and the underlying performance be, when we want to

- find *related rows* that have the same tag, even of the same entity?
- prevent *duplicate tags* being stored in a JSONB array column of distinct rows?
- find out which tags are most popular/related to one another?

## Practicality
> Special cases aren't special enough to break the rules.
> Although practicality beats purity.

The above arguments are not made from a stance of "relational purity". JSONB might not be a great option here for some pragmatic reasons, namely resource usage.

It seems that storing JSONB data may lead to [reduced query performance and increased database storage costs](https://heap.io/blog/engineering/when-to-avoid-jsonb-in-a-postgresql-schema) than would otherwise be incurred with a relational model. The storage costs are incurred due to likely indexing that would be applied to the column, while the performance penalty comes from lack of database statistics and perhaps even the complexity of querying an array of JSONB data, mentioned above.

## TLDR
It is understandable that we wanted to get a quick iteration on the prototype, so clients can be built around the API specification. Now it is time to consider refactoring the `tags` code to use a many-to-many model, so we don't have to migrate `tags` JSONB data to a relational model at a later point (e.g. in production).
